GSA-TTS/federal-website-standards
# Performance updates

## Context
This PR represents performance changes that started with [some awesome work](https://github.com/cloud-gov/pages-uswds-11ty/pull/76) by @ethangardner on the cloud.gov Eleventy template that the Federal Website Standards is based on.

## Description
I pulled in Ethan's changes which focus on shrinking the file size of the stylesheet by removing the images and fonts from the final stylesheet as well as not generating a sourcemap in production. From there I pulled in [PurgeCSS](https://purgecss.com/) as a plugin for PostCSS (the library that builds our stylesheet). PurgeCSS removes unused CSS selectors which drastically reduces the final size of the stylesheet, see below for some numbers. Finally I used Capo.js to evaluate the order of elements in the `<head>` tag and re-arrange. Having `<head>` tag elements in the proper order prevents several performance penalties, we were able to gain 3 points in the Lighthouse performance score on mobile from this change which was enough to move the score to 90/100 and to a green breakpoint instead of a yellow 87/100.

## Stylesheet file size numbers
In #258 I suggested that we start using USWDS packages as a way to reduce our CSS footprint. I evaluated switching to packages in addition to using PurgeCSS and found that the file size change didn't seem significant enough to warrant the developer/designer overhead that using packages would entail.

Original CSS stylesheet file size: 5.55MB
PurgeCSS CSS stylesheet file size: 84KB
PurgeCSS + USWDS Packages CSS stylesheet file size: 81KB

## How to verify this change
This one is a little hard to verify since it's behind the scenes and there's no visual or content difference to observe. A Lighthouse performance audit can be ran from Chrome developer tools against the live site and then against the preview URL and the preview URL should score significantly better.

Here is a GIF showing the difference in loading time on a desktop computer with cable internet connection between the live site homepage and the preview homepage: https://www.webpagetest.org/video/video.php?tests=241029_BiDc0P_97M,241029_AiDcR4_ABD&bg=ffffff&text=222222&end=visual&format=gif

## Additional information
Although this PR does not implement USWDS Packages I believe it fixes #258 by addressing the reason I wanted to use USWDS packages to begin with.

### Important PurgeCSS information
PurgeCSS works by removing unused CSS selectors and the style code associated with them. It is possible that some CSS selectors added in the future, either directly or by using additional USWDS components, might be deleted resulting in a component that isn't styled as expected. In this case the [PurgeCSS safelist](https://purgecss.com/safelisting.html) needs to be edited so that the selector(s) are preserved. Our PurgeCSS options/safelist lives in `config/purgecss-options.js`. I haven't observed this happening but this is just some FYSA for the future. I'll plan on creating an ADR directory to capture the decision to use PurgeCSS and include this note in that document as a risk.

🔗 [Preview URL](https://federalist-142078a8-f654-4daa-8f73-db9770b3ec7a.sites.pages.cloud.gov/preview/gsa-tts/federal-website-standards/caley/perf/)
